third_party/move/move-compiler-v2/src/pipeline/variable_coalescing.rs

```diff
@@ -0,0 +1,239 @@
+    /// See implementation comments for the sorting order.
+    fn sorted_live_interval_events(target: &FunctionTarget) -> Vec<LiveIntervalEvent> {
+        let live_intervals = Self::live_intervals(target);
+        let mut live_interval_events = vec![];
+        for (local, interval) in live_intervals.into_iter().enumerate() {
+            if let Some(interval) = interval {
+                live_interval_events.push(LiveIntervalEvent::Begin(
+                    local as TempIndex,
+                    interval.begin,
+                    (interval.end - interval.begin) as usize,
+                ));
+                live_interval_events.push(LiveIntervalEvent::End(local as TempIndex, interval.end));
+            }
+        }
+        live_interval_events.sort_by(|a, b| {
+            use LiveIntervalEvent::*;
+            match (a, b) {
+                // Sort events based on their code offsets (lower offset comes first).
+                _ if a.offset() < b.offset() => std::cmp::Ordering::Less,
+                _ if a.offset() > b.offset() => std::cmp::Ordering::Greater,
+                // If two events occur at the same offset, `End` comes before `Before`.
+                // This allows locals in `End` to be possibly remapped to locals in `Begin`
+                // at the same code offset.
+                (End(..), Begin(..)) => std::cmp::Ordering::Less,
+                (Begin(..), End(..)) => std::cmp::Ordering::Greater,
+                // If two locals `End` at the same offset, then we arbitrarily (but deterministically)
+                // use the local index to break the tie.
+                (End(local_a, _), End(local_b, _)) => local_a.cmp(local_b),
+                // If two locals `Begin` at the same offset, then we use the length of their live
+                // intervals to break the tie. The idea behind this heuristic is that remapping a
+                // local with shorter interval might make it available for remapping to other locals
+                // sooner. If the intervals are of the same length, we arbitrarily (but deterministically)
+                // use the local index to break the tie.
+                (Begin(local_a, _, length_a), Begin(local_b, _, length_b)) => {
+                    length_a.cmp(length_b).then_with(|| local_a.cmp(local_b))
+                },
+            }
+        });
+        live_interval_events
+    }
+
+    /// Compute the coalesceable locals of the given function target.
+    /// The result is a map, where for each mapping from local `t` to its coalesceable local `u`,
+    /// we can safely replace all occurrences of `t` with `u`.
+    /// This safety property follows from:
+    ///   - `t` and `u` are of the same type
+    ///   - either the live intervals of `t` and `u` do not overlap, in which case, they do not interfere
+    ///     with each others computations,
+    ///   - or `t` becomes live for the first time at the same code offset as `u` is last seen alive
```

Comment:
This is the same as "do not overlap" if you consider that a live range ending at a bytecode does not overlap with a live range beginning at that bytecode.  I see a comment above discussing that.  So you don't really need to be quite so wordy here.

Reply:
done